File: python/highest_scoring_word/highest_scoring_word.py
"""Given a string of words, you need to find the highest scoring word.

Each letter of a word scores points according to its position in the alphabet: a = 1, b = 2, c = 3 etc.

You need to return the highest scoring word as a string.

If two words score the same, return the word that appears earliest in the original string.

All letters will be lowercase and all inputs will be valid.
"""

def high(x):
    lista = x.split()
    suma, maximo = 0, 0
    for indice, palabra in enumerate(lista):
        for letra in palabra:
            suma += ord(letra)-96
        if (suma > maximo):
            maximo = suma
            indice_max = indice
        suma = 0
    return lista[indice_max]

print(high("aa b"))

"""
# Para estudiar:
def high(x):
    return max(x.split(), key=lambda k: sum(ord(c) - 96 for c in k))




def high(x):
    words=x.split(' ')
    list = []
    for i in words:
        scores = [sum([ord(char) - 96 for char in i])]
        list.append(scores)
    return words[list.index(max(list))]
    """

File: c++/highest_scoring_word/highest_scoring_word.cpp
/*Given a string of words, you need to find the highest scoring word.

	Each letter of a word scores points according to its position in the alphabet: a = 1, b = 2, c = 3 etc.

	You need to return the highest scoring word as a string.

	If two words score the same, return the word that appears earliest in the original string.

	All letters will be lowercase and all inputs will be valid.
	*/

#include <string>
#include <vector>
#include <iostream>
using namespace std;

std::string highestScoringWord(const std::string &str);

int main(void)
{
	std::string ans = highestScoringWord("what time are we climbing up the volcano");
	cout << ans;
};

std::string highestScoringWord(const std::string &str){
	std::string s = str + " ";
	std::vector<string> vec;
	std::string d = " ";
	size_t suma{}, maximo{}, j{}, elemento{}, last{}, next{};

	while ( ( next = s.find(d, last)) != std::string::npos ) {
		std::string palabra = s.substr(last, next - last);
		last = next + 1;
		vec.push_back(palabra);
		for (size_t i = 0; i < palabra.length(); i++ ) {
			suma += palabra[i];
		};
		if ( suma > maximo ) {
			maximo = suma;
			elemento = j;
		};
		j++;
		suma = 0;
	} ;
	return vec[elemento];
};


// Para estudiar:
/*
#include <string>

std::string highestScoringWord(const std::string &str)
{
  std::cout << str << std::endl;

  std::istringstream stm(str);
  std::string token, maxString;
  long maxValue = 0;

  auto GetValue = [](const std::string& s)
  {
    long r = 0;
    for (auto c : s)
      r += int(c) - int('a') + 1;

    return r;
  };

  while (std::getline(stm, token, ' '))
  {
    long actValue = GetValue(token);
    if (actValue > maxValue)
    {
      maxValue = actValue;
      maxString = token;
    }
  }

  return maxString;
}
*/
